toy-problems/generateParenthesis/generateParenthesis.js

```diff
@@ -0,0 +1,65 @@
+/* generate parenthesis
+given n pairs of parenthesis write
+function to generate all combinttion of well paired
+parenthesis
+
+Example:
+  Input: n = 3
+
+  Output: [
+    "((()))",
+    "(()())",
+    "(())()",
+    "()(())",
+    "()()()"
+  ]
+
+         (
+     (       )
+   (   )   (   )
+
+  '(((' 3
+  '(()' 1
+  '()' 0
+  '())' -1 BAD!!!
+
+  input: number of parens
+  output: is an array of strings(possible paren pairs)
+
+descision tree beccuase of permutations
+
+function generateParens that takes a number n
+  create counter variable set to 0
+  create result an array of valid paren combos
+
+  inner function parenBuilder
+
+  invoke parenBuilder with param
+
+  return result
+
+*/
+
+const generateParenthesis = (n) => {
+  const result = [];
+
+  const parenBuilder = (current, branchCount) => {
```

Comment:
`openParens` might be a better name for `branchCount` since we're really just counting the open parentheses.